### ISSUE TITLE
MockServletContext used by @WebAppConfiguration doesn't respect Spring Boot locations

### ISSUE THREAD

**Issue by ukimpact** [spring-projects/spring-boot#2654]
**Ver:** Liquibase 3.3.2; SB 1.2.2 
**Context:** Unit test with liquibase configuration and ```mockServletContext```
**Issue:** Exception message when adding @WebAppConfiguration to the test

Was able to reproduce by adding ```spring-boot-starter-web``` and the annotation to [sample project test](https://github.com/spring-projects/spring-boot/tree/master/spring-boot-samples/spring-boot-sample-liquibase)  

Simplified test for the sample project:

```java
@RunWith(SpringJUnit4ClassRunner.class)
@SpringApplicationConfiguration(classes = SampleLiquibaseApplication.class)
@WebAppConfiguration
public class SampleLiquibaseApplicationTests {
	@Test
	public void testDefaultSettings(){
    }
}
```

Non-blocker as the liquibase configuration actually succeeds. For the moment we are excluding liquibase auto configuration because of the message. Running application with real servlet context is fine.

The log:
```java
2015-03-14 18:35:16.040  WARN 58380 --- [           main] liquibase                                : Error initializing SpringLiquibase

java.io.FileNotFoundException: ServletContext resource [/] cannot be resolved to URL because it does not exist
	at org.springframework.web.context.support.ServletContextResource.getURL(ServletContextResource.java:156)
	at liquibase.integration.spring.SpringLiquibase$SpringResourceOpener.init(SpringLiquibase.java:77)
	at liquibase.resource.ClassLoaderResourceAccessor.<init>(ClassLoaderResourceAccessor.java:20)
	at liquibase.integration.spring.SpringLiquibase$SpringResourceOpener.<init>(SpringLiquibase.java:67)
	at liquibase.integration.spring.SpringLiquibase.createResourceOpener(SpringLiquibase.java:396)
	at liquibase.integration.spring.SpringLiquibase.createLiquibase(SpringLiquibase.java:357)
	at liquibase.integration.spring.SpringLiquibase.afterPropertiesSet(SpringLiquibase.java:315)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.invokeInitMethods(AbstractAutowireCapableBeanFactory.java:1625)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.initializeBean(AbstractAutowireCapableBeanFactory.java:1562)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:539)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:476)
	at org.springframework.beans.factory.support.AbstractBeanFactory$1.getObject(AbstractBeanFactory.java:303)
	at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:230)
	at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:299)
	at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:194)
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.preInstantiateSingletons(DefaultListableBeanFactory.java:755)
	at org.springframework.context.support.AbstractApplicationContext.finishBeanFactoryInitialization(AbstractApplicationContext.java:757)
	at org.springframework.context.support.AbstractApplicationContext.refresh(AbstractApplicationContext.java:480)
	at org.springframework.boot.SpringApplication.refresh(SpringApplication.java:689)
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:322)
	at org.springframework.boot.test.SpringApplicationContextLoader.loadContext(SpringApplicationContextLoader.java:100)
	at org.springframework.test.context.DefaultCacheAwareContextLoaderDelegate.loadContextInternal(DefaultCacheAwareContextLoaderDelegate.java:68)
	at org.springframework.test.context.DefaultCacheAwareContextLoaderDelegate.loadContext(DefaultCacheAwareContextLoaderDelegate.java:86)
	at org.springframework.test.context.DefaultTestContext.getApplicationContext(DefaultTestContext.java:72)
	at org.springframework.test.context.web.ServletTestExecutionListener.setUpRequestContextIfNecessary(ServletTestExecutionListener.java:170)
	at org.springframework.test.context.web.ServletTestExecutionListener.prepareTestInstance(ServletTestExecutionListener.java:110)
	at org.springframework.test.context.TestContextManager.prepareTestInstance(TestContextManager.java:212)
	at org.springframework.test.context.junit4.SpringJUnit4ClassRunner.createTest(SpringJUnit4ClassRunner.java:200)
	at org.springframework.test.context.junit4.SpringJUnit4ClassRunner$1.runReflectiveCall(SpringJUnit4ClassRunner.java:259)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.springframework.test.context.junit4.SpringJUnit4ClassRunner.methodBlock(SpringJUnit4ClassRunner.java:261)
	at org.springframework.test.context.junit4.SpringJUnit4ClassRunner.runChild(SpringJUnit4ClassRunner.java:219)
	at org.springframework.test.context.junit4.SpringJUnit4ClassRunner.runChild(SpringJUnit4ClassRunner.java:83)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.springframework.test.context.junit4.statements.RunBeforeTestClassCallbacks.evaluate(RunBeforeTestClassCallbacks.java:61)
	at org.springframework.test.context.junit4.statements.RunAfterTestClassCallbacks.evaluate(RunAfterTestClassCallbacks.java:68)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.springframework.test.context.junit4.SpringJUnit4ClassRunner.run(SpringJUnit4ClassRunner.java:163)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:74)
	at com.intellij.rt.execution.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:211)
	at com.intellij.rt.execution.junit.JUnitStarter.main(JUnitStarter.java:67)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:483)
	at com.intellij.rt.execution.application.AppMain.main(AppMain.java:134)

2015-03-14 18:35:16.133  INFO 58380 --- [           main] liquibase                                : Successfully acquired change log lock
2015-03-14 18:35:16.639  INFO 58380 --- [           main] liquibase                                : Creating database history table with name: PUBLIC.DATABASECHANGELOG
2015-03-14 18:35:16.644  INFO 58380 --- [           main] liquibase                                : Reading from PUBLIC.DATABASECHANGELOG
2015-03-14 18:35:16.652  INFO 58380 --- [           main] liquibase                                : classpath:/db/changelog/db.changelog-master.yaml: classpath:/db/changelog/db.changelog-master.yaml::1::marceloverdijk: Table person created
2015-03-14 18:35:16.652  INFO 58380 --- [           main] liquibase                                : classpath:/db/changelog/db.changelog-master.yaml: classpath:/db/changelog/db.changelog-master.yaml::1::marceloverdijk: ChangeSet classpath:/db/changelog/db.changelog-master.yaml::1::marceloverdijk ran successfully in 2ms
2015-03-14 18:35:16.660  INFO 58380 --- [           main] liquibase                                : classpath:/db/changelog/db.changelog-master.yaml: classpath:/db/changelog/db.changelog-master.yaml::2::marceloverdijk: New row inserted into person
2015-03-14 18:35:16.660  INFO 58380 --- [           main] liquibase                                : classpath:/db/changelog/db.changelog-master.yaml: classpath:/db/changelog/db.changelog-master.yaml::2::marceloverdijk: ChangeSet classpath:/db/changelog/db.changelog-master.yaml::2::marceloverdijk ran successfully in 1ms
2015-03-14 18:35:16.665  INFO 58380 --- [           main] liquibase                                : Successfully released change log lock

```

**Comment by philwebb** [comment]
I think I have a fix for this, but it's a little more involved that I'd like for 1.2.x. Since you currently just get a warning I'm going to target to 1.3.

**Comment by philwebb** [comment]
Incidentally, the root cause is that no `src/main/webapp` folder exists. As a work around you could create one or set a value in `@WebAppConfiguration` that points to `classpath:` folder that actually exists.

**Comment by ukimpact** [comment]
Thanks! I've tried the workaround already, and retried again just now to be sure - but I still get the same warning.

**Comment by philwebb** [comment]
I've backported part of the fix to 1.2.x. The problem was that file system resource such as `/src/main/webapp` weren't loaded because we were using the `DefaultResourceLoader` and not the `FileSystemResourceLoader`.

Creating an empty `src/main/webapp` folder should remove the warning with the latest snapshot. 1.3.0 will offer a complete fix.

**Comment by ukimpact** [comment]
Confirm the tests run clean with 1.2.3.BUILD-SNAPSHOT with empty webapp folder . Great stuff & thanks!
